### PR TITLE
Precompile Specific Packages

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Moves precompiles into their own isolated packages 

## How this works

Creates new packages for each precompile. Also splits the single file into `contract` and `config` files. Separates `stateful_precompile_test` to specific packages as well.

## How this was tested
Existing tests should cover this

### CODE REVIEW

File: precompile/allow_list.go [hunk 1]
```diff
@@ -48,10 +48,10 @@ type AllowListConfig struct {
 // the addresses in [AllowListAdmins].
 func (c *AllowListConfig) Configure(state StateDB, precompileAddr common.Address) error {
 	for _, enabledAddr := range c.EnabledAddresses {
-		setAllowListRole(state, precompileAddr, enabledAddr, AllowListEnabled)
+		SetAllowListRole(state, precompileAddr, enabledAddr, AllowListEnabled)
```
Comment (aaronbuchwald) [inline precompile/allow_list.go:51]: Should the allow list be its own package as well that can be imported to the other precompiles that use it?

File: precompile/deployerallowlist/config_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestVerifyContractDeployerConfig(t *testing.T) {
+	admins := []common.Address{{1}}
+	tests := []struct {
+		name          string
+		config        precompile.StatefulPrecompileConfig
+		expectedError string
+	}{
```
Comment (aaronbuchwald) [inline precompile/deployerallowlist/config_test.go:21]: Should we create a standard allow list test suite that can be shared across all of the precompiles that use it for both `Configure` and `Equal` ?

**Review by aaronbuchwald — APPROVED** [review]
LGTM

File: accounts/abi/bind/precompile_config_template.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// CUSTOM CODE STARTS HERE
+// Reference imports to suppress errors from unused imports. This code and any unnecessary imports can be removed.
+var (
+	_ = big.NewInt
+	_ = json.Unmarshal
+)
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply accounts/abi/bind/precompile_config_template.go]: removed

File: precompile/allow_list.go (hunk 1 above)
Comment (ceyonur) [reply precompile/allow_list.go:52]: i left it out since it was not an actual precompile, but yeah we can move it.

File: plugin/evm/vm_test.go [hunk 4]
```diff
@@ -2424,8 +2430,8 @@ func TestFeeManagerChangeFee(t *testing.T) {
 	testHighFeeConfig := testLowFeeConfig
 	testHighFeeConfig.MinBaseFee = big.NewInt(28_000_000_000)
 
-	data, err := precompile.PackSetFeeConfig(testHighFeeConfig)
-	require.NoError(t, err)
+	data, err := feemanager.PackSetFeeConfig(testHighFeeConfig)
+	assert.NoError(t, err)
```
Comment (darioush) [inline plugin/evm/vm_test.go:2434]: nit: could we leave this as `require`?

**Review by darioush — APPROVED** [review]
Good improvements, just nits.

File: precompile/deployerallowlist/config_test.go (hunk 2 above)
Comment (ceyonur) [reply precompile/deployerallowlist/config_test.go:21]: I tried to find a simple way but we need to give concrete struct as the input to the suit. I prefer unit tests to be dumb as possible.

File: precompile/rewardmanager/config.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+func (i *InitialRewardConfig) Verify() error {
+	switch {
+	case i.AllowFeeRecipients && i.RewardAddress != (common.Address{}):
+		return ErrCannotEnableBothRewards
+	default:
+		return nil
+	}
+}
+
+func (c *InitialRewardConfig) Equal(other *InitialRewardConfig) bool {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/rewardmanager/config.go:34]: yup good catch!

File: precompile/nativeminter/config_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+				map[common.Address]*math.HexOrDecimal256{
+					common.HexToAddress("0x01"): math.NewHexOrDecimal256(1),
+				}),
+			other: NewContractNativeMinterConfig(big.NewInt(3), admins, nil,
+				map[common.Address]*math.HexOrDecimal256{
+					common.HexToAddress("0x02"): math.NewHexOrDecimal256(1),
+				}),
+			expected: false,
+		},
+
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/nativeminter/config_test.go:129]: removed

File: plugin/evm/vm_test.go (hunk 4 above)
Comment (ceyonur) [reply plugin/evm/vm_test.go:2434]: sure, seems like a bad merge conflict.

**Comment by gatesyp** [comment]
Any thoughts on using Caddy as inspiration for the plugin system? this PR seems more mature, so OK if not

Caddy has a great plugin system that would be perfect for precompiles (imo). Cant use it directly but could model something on it

**Comment by ceyonur** [comment]
Hi @gatesyp,
Yes we currently working on a draft to introduce precompiles as modules on top of this PR. You can check it out: https://github.com/ava-labs/subnet-evm/pull/426/

**Pull request merged**